File: code/packages/ruby/brainfuck/test/test_lexer.rb
# frozen_string_literal: true

require_relative "test_helper"

# ================================================================
# Tests for the Brainfuck Lexer
# ================================================================
#
# These tests verify that the grammar-driven lexer, when loaded
# with brainfuck.tokens, correctly tokenizes Brainfuck source.
#
# Brainfuck tokenization has two important properties that we test:
#
#   1. COMMAND recognition — the 8 command characters (> < + - . , [ ])
#      must each produce the correct token type and value.
#
#   2. COMMENT skipping — every non-command character is silently
#      consumed. Letters, digits, spaces, punctuation are all comments.
#      They MUST NOT appear in the token stream.
#
# We are not testing the GrammarLexer engine itself (tested in the
# lexer gem) — we are testing that brainfuck.tokens correctly
# describes Brainfuck's lexical rules.
# ================================================================

class TestBrainfuckLexer < Minitest::Test
  TT = CodingAdventures::Lexer::TokenType

  # Brainfuck token types are plain strings (not in TokenType::ALL)
  RIGHT_TYPE      = "RIGHT"
  LEFT_TYPE       = "LEFT"
  INC_TYPE        = "INC"
  DEC_TYPE        = "DEC"
  OUTPUT_TYPE     = "OUTPUT"
  INPUT_TYPE      = "INPUT"
  LOOP_START_TYPE = "LOOP_START"
  LOOP_END_TYPE   = "LOOP_END"

  # ------------------------------------------------------------------
  # Helpers
  # ------------------------------------------------------------------

  def tokenize(source)
    CodingAdventures::Brainfuck::Lexer.tokenize(source)
  end

  def command_tokens(source)
    # Filter out EOF so tests can just check command tokens
    tokenize(source).reject { |t| t.type == TT::EOF }
  end

  def token_types(source)
    command_tokens(source).map(&:type)
  end

  # ------------------------------------------------------------------
  # Grammar path sanity check
  # ------------------------------------------------------------------
  # The lexer must be able to find brainfuck.tokens at the path
  # computed by the module constant. If this test fails, the directory
  # navigation arithmetic in GRAMMAR_DIR is wrong.

  def test_grammar_path_exists
    assert File.exist?(CodingAdventures::Brainfuck::Lexer::BF_TOKENS_PATH),
      "brainfuck.tokens should exist at #{CodingAdventures::Brainfuck::Lexer::BF_TOKENS_PATH}"
  end

  # ------------------------------------------------------------------
  # Individual command tokens
  # ------------------------------------------------------------------
  # Each of the 8 Brainfuck commands must be recognized as its own
  # token type. The value field must contain the exact source character.

  def test_right_token
    # ">" moves the data pointer one cell to the right.
    tokens = command_tokens(">")
    assert_equal 1, tokens.length
    assert_equal RIGHT_TYPE, tokens[0].type
    assert_equal ">", tokens[0].value
  end

  def test_left_token
    # "<" moves the data pointer one cell to the left.
    tokens = command_tokens("<")
    assert_equal 1, tokens.length
    assert_equal LEFT_TYPE, tokens[0].type
    assert_equal "<", tokens[0].value
  end

  def test_inc_token
    # "+" increments the byte at the current cell.
    tokens = command_tokens("+")
    assert_equal 1, tokens.length
    assert_equal INC_TYPE, tokens[0].type
    assert_equal "+", tokens[0].value
  end

  def test_dec_token
    # "-" decrements the byte at the current cell.
    tokens = command_tokens("-")
    assert_equal 1, tokens.length
    assert_equal DEC_TYPE, tokens[0].type
    assert_equal "-", tokens[0].value
  end

  def test_output_token
    # "." writes the current cell's byte value to stdout.
    tokens = command_tokens(".")
    assert_equal 1, tokens.length
    assert_equal OUTPUT_TYPE, tokens[0].type
    assert_equal ".", tokens[0].value
  end

  def test_input_token
    # "," reads one byte from stdin into the current cell.
    tokens = command_tokens(",")
    assert_equal 1, tokens.length
    assert_equal INPUT_TYPE, tokens[0].type
    assert_equal ",", tokens[0].value
  end

  def test_loop_start_token
    # "[" jumps forward past matching "]" if current cell is zero.
    tokens = command_tokens("[")
    assert_equal 1, tokens.length
    assert_equal LOOP_START_TYPE, tokens[0].type
    assert_equal "[", tokens[0].value
  end

  def test_loop_end_token
    # "]" jumps back to matching "[" if current cell is nonzero.
    tokens = command_tokens("]")
    assert_equal 1, tokens.length
    assert_equal LOOP_END_TYPE, tokens[0].type
    assert_equal "]", tokens[0].value
  end

  # ------------------------------------------------------------------
  # Comment skipping
  # ------------------------------------------------------------------
  # All non-command characters are comments. The lexer's skip: section
  # consumes them silently. No comment tokens should appear in output.

  def test_letters_are_comments
    # Alphabetic characters often appear as prose comments in Brainfuck.
    # "hello" produces zero command tokens.
    types = token_types("hello world")
    assert_empty types, "Letter comments should not produce any tokens"
  end

  def test_digits_are_comments
    # Numeric annotations (like "cell 0") are comments too.
    types = token_types("123 cell value")
    assert_empty types, "Digit comments should not produce any tokens"
  end

  def test_comments_between_commands
    # A common Brainfuck convention is to annotate commands inline:
    #   +  increment cell 0
    #   >  move right to cell 1
    # The lexer should strip the annotation and keep only + and >.
    types = token_types("+ increment cell 0 > move right")
    assert_equal [INC_TYPE, RIGHT_TYPE], types
  end

  def test_mixed_whitespace_and_commands
    # Whitespace between commands (spaces, tabs, newlines) is skipped.
    # A well-formatted Brainfuck program uses indentation and newlines.
    source = "  +  \n  >  \n  -  "
    types = token_types(source)
    assert_equal [INC_TYPE, RIGHT_TYPE, DEC_TYPE], types
  end

  def test_all_commands_in_sequence
    # All eight commands together, no comments. Order must be preserved.
    source = "><+-.,[]"
    types = token_types(source)
    expected = [
      RIGHT_TYPE, LEFT_TYPE, INC_TYPE, DEC_TYPE,
      OUTPUT_TYPE, INPUT_TYPE, LOOP_START_TYPE, LOOP_END_TYPE
    ]
    assert_equal expected, types
  end

  # ------------------------------------------------------------------
  # Empty source
  # ------------------------------------------------------------------
  # An empty Brainfuck program is valid. The only token should be EOF.

  def test_empty_source_returns_eof
    tokens = tokenize("")
    assert_equal 1, tokens.length, "Empty source should produce exactly one token (EOF)"
    assert_equal TT::EOF, tokens[0].type
  end

  def test_only_comments_returns_eof
    # A source with nothing but comments produces an empty command stream.
    tokens = tokenize("This is a Brainfuck program that does nothing.")
    assert_equal TT::EOF, tokens.last.type
    non_eof = tokens.reject { |t| t.type == TT::EOF }
    assert_empty non_eof, "Only-comment source should produce no command tokens"
  end

  # ------------------------------------------------------------------
  # Line and column tracking
  # ------------------------------------------------------------------
  # The lexer must track line and column numbers accurately so that
  # error messages from the parser can report meaningful locations.
  # Line numbers start at 1, column numbers start at 1.

  def test_line_tracking_across_newlines
    # A newline in a comment should advance the line counter.
    # The ">" on line 3 (after two lines of comments) should report line 3.
    source = "comment line 1\ncomment line 2\n>"
    tokens = command_tokens(source)
    assert_equal 1, tokens.length
    assert_equal RIGHT_TYPE, tokens[0].type
    assert_equal 3, tokens[0].line, "Token after two newlines should be on line 3"
  end

  def test_column_tracking_on_first_line
    # Spaces before the first command push the column to the right.
    # Four spaces before "+" means column 5.
    tokens = command_tokens("    +")
    assert_equal 1, tokens.length
    assert_equal 5, tokens[0].col, "Token after 4 spaces should be at column 5"
  end

  def test_eof_always_present
    # The token stream always ends with EOF regardless of source content.
    ["", "+", "++", "hello", "++[>+<-]"].each do |src|
      tokens = tokenize(src)
      assert_equal TT::EOF, tokens.last.type, "EOF must be last token for #{src.inspect}"
    end
  end

  # ------------------------------------------------------------------
  # Canonical Brainfuck example: ++[>+<-]
  # ------------------------------------------------------------------
  # This program is the simplest non-trivial Brainfuck idiom.
  # It adds the value of cell 0 to cell 1 (2 + 0 = 2 in cell 1),
  # leaving cell 0 at 0. It exercises all the pointer, arithmetic,
  # and loop control tokens together.
  #
  #   ++      set cell 0 to 2
  #   [       loop while cell 0 != 0
  #     >+    move right, increment cell 1
  #     <-    move left, decrement cell 0
  #   ]       end loop (exits when cell 0 = 0)

  def test_canonical_plus_plus_loop
    source = "++[>+<-]"
    types = token_types(source)
    expected = [
      INC_TYPE, INC_TYPE,           # ++
      LOOP_START_TYPE,              # [
        RIGHT_TYPE, INC_TYPE,       # >+
        LEFT_TYPE, DEC_TYPE,        # <-
      LOOP_END_TYPE                 # ]
    ]
    assert_equal expected, types
  end

  def test_canonical_plus_plus_loop_values
    # Verify that the value field contains the exact source character.
    source = "++[>+<-]"
    tokens = command_tokens(source)
    values = tokens.map(&:value)
    assert_equal ["+", "+", "[", ">", "+", "<", "-", "]"], values
  end

  def test_canonical_with_inline_comments
    # The same program annotated with comments should produce identical tokens.
    # This is how Brainfuck programs are documented in the wild.
    source = "++ add two [>+ move+inc <-  decrement  ] end loop"
    types = token_types(source)
    expected = [
      INC_TYPE, INC_TYPE,
      LOOP_START_TYPE,
        RIGHT_TYPE, INC_TYPE,
        LEFT_TYPE, DEC_TYPE,
      LOOP_END_TYPE
    ]
    assert_equal expected, types
  end
end

File: code/packages/python/brainfuck/tests/test_lexer.py
"""Tests for the Brainfuck lexer thin wrapper.

These tests verify that the grammar-driven lexer, configured with
``brainfuck.tokens``, correctly tokenizes Brainfuck source text.

Brainfuck is one of the simplest languages to lex: exactly 8 single-character
command tokens, with everything else silently discarded. There are no strings,
no numbers, no keywords, and no multi-character tokens.

Test Categories
---------------

1. All 8 command tokens — each of ><+-.,[] produces the correct token type
2. Comment skipping — non-command characters are silently discarded
3. Position tracking — line and column numbers advance correctly
4. Empty source — empty string produces only EOF
5. Canonical ``++[>+<-]`` program — the classic loop pattern
6. Mixed commands and comments — realistic annotated Brainfuck source
7. Multi-line programs — newlines are skipped, line counter advances
8. EOF is always the last token
"""

from __future__ import annotations

import pytest

from brainfuck.lexer import create_brainfuck_lexer, tokenize_brainfuck
from lexer import Token


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def token_types(source: str) -> list[str]:
    """Tokenize and return just the type names, excluding EOF."""
    tokens = tokenize_brainfuck(source)
    return [
        t.type if isinstance(t.type, str) else t.type.name
        for t in tokens
        if (t.type if isinstance(t.type, str) else t.type.name) != "EOF"
    ]


def token_values(source: str) -> list[str]:
    """Tokenize and return just the values, excluding EOF."""
    tokens = tokenize_brainfuck(source)
    return [
        t.value
        for t in tokens
        if (t.type if isinstance(t.type, str) else t.type.name) != "EOF"
    ]


# ---------------------------------------------------------------------------
# All 8 command tokens
# ---------------------------------------------------------------------------


class TestAllEightCommands:
    """Verify that each of the 8 Brainfuck command characters produces the correct token."""

    def test_right_command(self) -> None:
        """The ``>`` character produces a RIGHT token.

        ``>`` moves the data pointer one cell to the right on the tape.
        Starting from cell 0, after ``>``, the pointer is at cell 1.
        """
        tokens = tokenize_brainfuck(">")
        first = tokens[0]
        type_name = first.type if isinstance(first.type, str) else first.type.name
        assert type_name == "RIGHT"
        assert first.value == ">"

    def test_left_command(self) -> None:
        """The ``<`` character produces a LEFT token.

        ``<`` moves the data pointer one cell to the left on the tape.
        """
        tokens = tokenize_brainfuck("<")
        first = tokens[0]
        type_name = first.type if isinstance(first.type, str) else first.type.name
        assert type_name == "LEFT"
        assert first.value == "<"

    def test_inc_command(self) -> None:
        """The ``+`` character produces an INC token.

        ``+`` increments the byte at the current data pointer. After 256
        increments from 0, the cell wraps back to 0 (unsigned byte arithmetic).
        """
        tokens = tokenize_brainfuck("+")
        first = tokens[0]
        type_name = first.type if isinstance(first.type, str) else first.type.name
        assert type_name == "INC"
        assert first.value == "+"

    def test_dec_command(self) -> None:
        """The ``-`` character produces a DEC token.

        ``-`` decrements the byte at the current data pointer. From 0 it
        wraps to 255.
        """
        tokens = tokenize_brainfuck("-")
        first = tokens[0]
        type_name = first.type if isinstance(first.type, str) else first.type.name
        assert type_name == "DEC"
        assert first.value == "-"

    def test_output_command(self) -> None:
        """The ``.`` character produces an OUTPUT token.

        ``.`` outputs the byte at the current data pointer as an ASCII character.
        Cell value 65 outputs the letter 'A'.
        """
        tokens = tokenize_brainfuck(".")
        first = tokens[0]
        type_name = first.type if isinstance(first.type, str) else first.type.name
        assert type_name == "OUTPUT"
        assert first.value == "."

    def test_input_command(self) -> None:
        """The ``,`` character produces an INPUT token.

        ``,`` reads one byte from the input stream into the current cell.
        """
        tokens = tokenize_brainfuck(",")
        first = tokens[0]
        type_name = first.type if isinstance(first.type, str) else first.type.name
        assert type_name == "INPUT"
        assert first.value == ","

    def test_loop_start_command(self) -> None:
        """The ``[`` character produces a LOOP_START token.

        ``[`` begins a loop. If the current cell is zero, execution jumps
        past the matching ``]``. The lexer tokenizes it regardless of whether
        a matching ``]`` exists — bracket matching is the parser's job.
        """
        tokens = tokenize_brainfuck("[")
        first = tokens[0]
        type_name = first.type if isinstance(first.type, str) else first.type.name
        assert type_name == "LOOP_START"
        assert first.value == "["

    def test_loop_end_command(self) -> None:
        """The ``]`` character produces a LOOP_END token.

        ``]`` ends a loop. If the current cell is nonzero, execution jumps
        back to the matching ``[``. The idiom ``[-]`` clears the current cell.
        """
        tokens = tokenize_brainfuck("]")
        first = tokens[0]
        type_name = first.type if isinstance(first.type, str) else first.type.name
        assert type_name == "LOOP_END"
        assert first.value == "]"


# ---------------------------------------------------------------------------
# Comment skipping
# ---------------------------------------------------------------------------


class TestCommentSkipping:
    """Verify that non-command characters are silently discarded."""

    def test_alphabetic_comment_produces_no_tokens(self) -> None:
        """Alphabetic text with no commands produces an empty token list.

        In Brainfuck, any character that isn't one of the 8 commands is a
        comment. The COMMENT skip pattern in brainfuck.tokens matches all
        non-command, non-whitespace characters.
        """
        types = token_types("this is a comment")
        assert types == []

    def test_commands_with_trailing_comment(self) -> None:
        """Commands followed by a comment produce only the command tokens.

        Brainfuck programmers commonly write annotations like:
            +++ set cell to 3
        The text "set cell to 3" should be silently discarded.
        """
        types = token_types("+++ set cell to 3")
        assert types == ["INC", "INC", "INC"]

    def test_digits_are_comments(self) -> None:
        """Digit characters are not Brainfuck commands.

        ``42`` contains no commands, just two comment characters (digits).
        This is a common misconception: ``+ 10 times`` does NOT run ``+``
        ten times — the ``10`` is just prose comment text.
        """
        types = token_types("42")
        assert types == []

    def test_commands_interspersed_with_comments(self) -> None:
        """Commands amid prose: only the command characters survive.

        ``> move right < move left`` should produce RIGHT, LEFT — the
        surrounding words are comments.
        """
        types = token_types("> move right < move left")
        assert types == ["RIGHT", "LEFT"]


# ---------------------------------------------------------------------------
# Empty source
# ---------------------------------------------------------------------------


class TestEmptySource:
    """Verify behaviour on empty and whitespace-only sources."""

    def test_empty_string_produces_only_eof(self) -> None:
        """An empty source string returns exactly one token: EOF.

        An empty Brainfuck program is valid — it simply does nothing.
        The lexer always appends a synthetic EOF sentinel token.
        """
        tokens = tokenize_brainfuck("")
        assert len(tokens) == 1
        type_name = tokens[0].type if isinstance(tokens[0].type, str) else tokens[0].type.name
        assert type_name == "EOF"

    def test_whitespace_only_produces_only_eof(self) -> None:
        """A source containing only whitespace returns exactly one token: EOF.

        All whitespace is consumed by the WHITESPACE skip pattern, leaving
        no command tokens. The result is the same as an empty source.
        """
        tokens = tokenize_brainfuck("   \\t\\n   ")
        assert len(tokens) == 1
        type_name = tokens[0].type if isinstance(tokens[0].type, str) else tokens[0].type.name
        assert type_name == "EOF"


# ---------------------------------------------------------------------------
# EOF is always last
# ---------------------------------------------------------------------------


class TestEOFAlwaysLast:
    """Verify that the last token in every result is EOF."""

    def test_eof_is_last_for_single_command(self) -> None:
        """A single command produces [command, EOF].

        The EOF token is a synthetic sentinel that the generic lexer always
        appends. It tells the parser there are no more tokens.
        """
        tokens = tokenize_brainfuck("+")
        last = tokens[-1]
        type_name = last.type if isinstance(last.type, str) else last.type.name
        assert type_name == "EOF"

    def test_eof_is_last_for_multiple_commands(self) -> None:
        """Multiple commands produce [cmd, cmd, ..., EOF].

        Regardless of how many command tokens the source produces, the
        final token is always EOF.
        """
        tokens = tokenize_brainfuck("++--")
        last = tokens[-1]
        type_name = last.type if isinstance(last.type, str) else last.type.name
        assert type_name == "EOF"


# ---------------------------------------------------------------------------
# Position tracking
# ---------------------------------------------------------------------------


class TestPositionTracking:
    """Verify that line and column numbers are tracked correctly."""

    def test_first_command_is_at_line_1_column_1(self) -> None:
        """The first command in a source is at line 1, column 1.

        Both line and column are 1-indexed (first line is 1, first column
        is 1). This follows the Token interface convention.
        """
        tokens = tokenize_brainfuck(">")
        assert tokens[0].line == 1
        assert tokens[0].column == 1

    def test_column_advances_along_first_line(self) -> None:
        """Commands on the same line have increasing column numbers.

        For ``><``, the ``>`` is at column 1 and ``<`` is at column 2.
        """
        tokens = tokenize_brainfuck("><")
        assert tokens[0].column == 1
        assert tokens[1].column == 2

    def test_line_advances_across_newlines(self) -> None:
        """A newline causes the line counter to increment.

        The WHITESPACE skip pattern consumes newlines, which causes the
        lexer engine to advance its line counter. A command on the second
        line should have line == 2.

        Source::

            +   <- line 1
            -   <- line 2
        """
        tokens = tokenize_brainfuck("+\n-")
        assert tokens[0].line == 1
        assert tokens[1].line == 2

    def test_column_after_spaces(self) -> None:
        """Spaces advance the column counter.

        For ``+   -``, the ``+`` is at column 1 and ``-`` is at column 5
        (after 3 spaces).
        """
        tokens = tokenize_brainfuck("+   -")
        assert tokens[0].column == 1
        assert tokens[1].column == 5


# ---------------------------------------------------------------------------
# Canonical ++[>+<-] pattern
# ---------------------------------------------------------------------------


class TestCanonicalPattern:
    """Tests for the canonical ``++[>+<-]`` Brainfuck idiom."""

    def test_produces_eight_command_tokens_plus_eof(self) -> None:
        """``++[>+<-]`` produces exactly 8 command tokens + EOF = 9 total.

        The sequence encodes 8 Brainfuck commands:
            ++     — set cell 0 to 2
            [      — loop while cell 0 is nonzero
              >    — move to cell 1
              +    — increment cell 1
              <    — return to cell 0
              -    — decrement cell 0
            ]      — end loop

        Result: cell 0 = 0, cell 1 = 2.
        """
        tokens = tokenize_brainfuck("++[>+<-]")
        assert len(tokens) == 9  # 8 commands + EOF

    def test_produces_correct_type_sequence(self) -> None:
        """``++[>+<-]`` produces the exact expected token type sequence."""
        types = token_types("++[>+<-]")
        assert types == [
            "INC",         # +
            "INC",         # +
            "LOOP_START",  # [
            "RIGHT",       # >
            "INC",         # +
            "LEFT",        # <
            "DEC",         # -
            "LOOP_END",    # ]
        ]

    def test_produces_correct_values(self) -> None:
        """``++[>+<-]`` produces the exact expected token value sequence."""
        values = token_values("++[>+<-]")
        assert values == ["+", "+", "[", ">", "+", "<", "-", "]"]

    def test_with_inline_comments_same_as_without(self) -> None:
        """Adding comments to ``++[>+<-]`` does not change the token sequence.

        A Brainfuck programmer might write:
            ++ setup  [ copy loop >+ cell1 <-  ]

        The comments ``setup``, ``copy loop``, ``cell1`` should be discarded,
        leaving the same 8 command tokens.
        """
        clean_types = token_types("++[>+<-]")
        commented_types = token_types("++ setup  [ copy loop >+ cell1 <-  ]")
        assert clean_types == commented_types

    def test_all_eight_distinct_commands(self) -> None:
        """``><+-.,[]`` contains one of each Brainfuck command.

        This verifies that all 8 token types are distinct and correctly
        identified by the lexer.
        """
        types = token_types("><+-.,[]")
        assert types == [
            "RIGHT", "LEFT", "INC", "DEC", "OUTPUT", "INPUT",
            "LOOP_START", "LOOP_END",
        ]


# ---------------------------------------------------------------------------
# create_brainfuck_lexer factory
# ---------------------------------------------------------------------------


class TestCreateBrainfuckLexer:
    """Verify that the ``create_brainfuck_lexer`` factory function works."""

    def test_factory_returns_grammar_lexer(self) -> None:
        """``create_brainfuck_lexer`` returns a ``GrammarLexer`` instance.

        The factory function is provided for callers who want to inspect
        the grammar or control tokenization step-by-step.
        """
        from lexer import GrammarLexer

        lexer = create_brainfuck_lexer("+")
        assert isinstance(lexer, GrammarLexer)

    def test_factory_tokenize_matches_convenience_function(self) -> None:
        """``create_brainfuck_lexer(s).tokenize()`` equals ``tokenize_brainfuck(s)``.

        Both code paths should produce identical token lists.
        """
        source = "++[>+<-]"
        via_factory = create_brainfuck_lexer(source).tokenize()
        via_convenience = tokenize_brainfuck(source)

        factory_types = [
            t.type if isinstance(t.type, str) else t.type.name
            for t in via_factory
        ]
        convenience_types = [
            t.type if isinstance(t.type, str) else t.type.name
            for t in via_convenience
        ]
        assert factory_types == convenience_types
